### PR TITLE
Rails admin

### PR DESCRIPTION
**Pull request by slawosz**
Quick PoC for admin panel. Currently allows changing password and other user attributes.
Requires new role for the user, `superadmin`:
```
user = User.find(...)
user.user_roles.create name: 'superadmin'
```

To make it mergeable:
- [x] make sure it is accessible only via authentication

Screenshot from 2020-03-31 17-45-07

Screenshot from 2020-03-31 17-45-27

### CODE REVIEW

File: psd-web/config/initializers/rails_admin.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  config.authenticate_with do
+    warden.authenticate! scope: :user
+  end
+  config.authorize_with do |_controller|
+    if warden.session(:user)[TwoFactorAuthentication::NEED_AUTHENTICATION]
+      redirect_to "/two-factor"
+    end
+    if !current_user.is_superadmin?
+      redirect_to main_app.root_path
+    end
```
Comment (scruti) [inline psd-web/config/initializers/rails_admin.rb:91]: Suggestion:
Could we invert the checks so we ensure the user `is_superadmin?` before sending them to 2FA page if needed?
It seems like the user will be asked for 2FA (when needed by its session) even if is not a super admin. What would be the point?

Comment (slawosz) [reply psd-web/config/initializers/rails_admin.rb:91]: Good point. I will change it. However it doesn't matter as the user in such case will end up in the same page.

File: psd-web/app/controllers/application_controller.rb [hunk 2]
```diff
@@ -15,6 +15,13 @@ class ApplicationController < ActionController::Base
 
   helper_method :nav_items, :secondary_nav_items, :previous_search_params, :current_user
 
+
+  def current_admin
+    if current_user && current_user.is_superadmin?
+      current_user
+    end
+  end
```
Comment (scruti) [inline psd-web/app/controllers/application_controller.rb:23]: Do we use this anywhere or need it at all?

File: psd-web/config/initializers/rails_admin.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  config.included_models = %w[User]
+end
+
+module RailsAdminTwoFactor
+  def user_two_factor_authentication_path
+    "/two-factor"
+  end
+end
+
+RailsAdmin::MainController.include RailsAdminTwoFactor
```
Comment (scruti) [inline psd-web/config/initializers/rails_admin.rb:103]: A little bit confused about why are we declaring this here but at the same time we are manually redirecting to `/two-factor`  inside `config.authorize_with` block.

Little bit of context? :)

File: psd-web/spec/features/admin_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+require "rails_helper"
+
+RSpec.feature "Admin Panel", :with_elasticsearch, :with_stubbed_mailer, :with_stubbed_notify, type: :feature do
+  include ActiveSupport::Testing::TimeHelpers
+
+  let(:investigation) { create(:project) }
+  let(:user) { create(:user, :activated, has_viewed_introduction: true) }
+  let(:password) { "2538fhdkvuULE36f" }
```
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:8]: Please change: 
I believe this is the same as the password for user factory. Being that the case we can avoid this explicit declaration and just use `user.password` in `fill_in_credentials` as we do with `user.email`.

File: psd-web/spec/features/admin_spec.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  def fill_in_credentials(password_override: nil)
+    fill_in "Email address", with: user.email
+    if password_override
+      fill_in "Password", with: password_override
+    else
+      fill_in "Password", with: password
+    end
+    click_on "Continue"
+  end
```
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:18]: Suggestion:
Being the same as the `sign-in` feature spec... time to extract and share the helper?
`def fill_in_credentials(user, password_override: nil)` could be a generic interface that allows us to fill any sign-in form.

File: psd-web/spec/features/admin_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+  context "when user tries to access admin", :with_2fa do
+    it "needs to sign in" do
```
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:21]: Suggestion:
Move to `feature` / `scenario` syntax. Did we reach any agreement on this? :thinking: 
Trying to normalise what syntax we use for feature tests.

File: psd-web/spec/features/admin_spec.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      visit "/admin"
+      expect(page).to have_css("h1", text: "Sign in")
+      visit "/sign-in"
```
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:24]: :scissors: 
We are already in the `Sign in page`. It doesn't seem we need to do a further request here, right?

File: psd-web/spec/features/admin_spec.rb (hunk 4 above)
Comment (slawosz) [reply psd-web/spec/features/admin_spec.rb:8]: But the thing is, you don't have access to `user.password`, as its write only attribute.

File: psd-web/spec/features/admin_spec.rb (hunk 6 above)
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:21]:
```suggestion
    it "needs to be signed in as super admin" do
```

File: psd-web/spec/features/admin_spec.rb (hunk 4 above)
Comment (scruti) [reply psd-web/spec/features/admin_spec.rb:8]: We do in the factory generated user: 
```
 user.password
=> "2538fhdkvuULE36f"
```

File: psd-web/config/initializers/rails_admin.rb (hunk 1 above)
Comment (scruti) [reply psd-web/config/initializers/rails_admin.rb:91]: Yeah is more about the user experience turning weird :D

File: psd-web/spec/features/admin_spec.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      fill_in_credentials
```
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:25]: Suggestion:
add expectation about not being on the admin panel after submitting credentials.

File: psd-web/config/initializers/rails_admin.rb [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    end
+  end
+
+  config.included_models = %w[User]
+end
+
+# Due two_factor_authentication_gem implementation, which is being hooked to
+# RailsAdmin controller, we need to define this path.
+module RailsAdminTwoFactor
+  # This method will be called inside RailsAdmin, but is only defined for host app.
```
Comment (scruti) [inline psd-web/config/initializers/rails_admin.rb:100]: Thanks for clarifying this!

File: psd-web/spec/features/admin_spec.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+require "rails_helper"
+
+RSpec.feature "Admin Panel", :with_elasticsearch, :with_stubbed_mailer, :with_stubbed_notify, type: :feature do
+  include ActiveSupport::Testing::TimeHelpers
```
Comment (scruti) [inline psd-web/spec/features/admin_spec.rb:4]: :scissors: I don't think we are using any of the helpers right?

**Comment by nasirkhanbeis** [comment]
Re-tested this in review app with permission as super user, I can manage user by being able to change phone  number and name, all good. 
At moment review app is on new db so you  will see you only fewer users.

**Pull request merged**